### ISSUE TITLE
InvoiceLineItemCollection provided by getLines doesn't get same auth request options with autoPagingIterable

### ISSUE THREAD

**Issue by keithblaha** [stripe/stripe-java#264]
I'm using this library in a multithreaded environment against various keys granted via oauth by Stripe users for my app, so I provide the appropriate key in request options for each call I make. However, it seems like some of the models' methods don't copy these credentials on additional requests that are made.

e.g., I call Invoice.list passing my request options, and then on each invoice I call getLines and use the autoPagingIterable to access all invoice line items. However, for any invoice with enough line items such that an additional request is needed I get an exception like this:

```
java.lang.RuntimeException: Unable to lazy-load stripe objects
Caused by: com.stripe.exception.AuthenticationException: No API key provided. (HINT: set your API key using 'Stripe.apiKey = <API-KEY>'. You can generate API keys from the Stripe web interface. See https://stripe.com/api for details or email support@stripe.com if you have questions.
```

e: I'm using 1.41.0

**Comment by keithblaha** [comment]
looks related to #237

**Comment by brandur** [comment]
@keithblaha Hey! Sorry about the trouble here, and thanks for logging this.

So the auto-paging iterator does theoretically pass through an API key along with any other request options that might be have been included with the original list request:

https://github.com/stripe/stripe-java/blob/89acae5f968a8e72688e59b8e5a40068143fcbc1/src/main/java/com/stripe/model/PagingIterator.java#L55

And we do have some basic tests in place to verify this behavior:

https://github.com/stripe/stripe-java/blob/89acae5f968a8e72688e59b8e5a40068143fcbc1/src/test/java/com/stripe/model/PagingIteratorTest.java#L84,L105

It's quite possible that there is some case that we haven't considered though. Would it be possible for you to assemble a basic repro case that demonstrates (in code) how you're using these classes? That would help us a lot in getting to the bottom of this.

**Comment by keithblaha** [comment]
Thanks for responding so fast @brandur! Sorry my title/original message we're all sloppy and probably hard to parse- I was just filing this quickly while in the middle of some stuff. I want to reiterate that this issue is particular to running autopagination within an already occurring autopagination (in this case, off of line items for an invoice)

I've created a new stripe account to test this against. It just has 1 customer with 1 invoice which has many line items on it (enough to cause an additional request). This gist successfully duplicates my issue using that account: https://gist.github.com/keithblaha/018594d285b1568792d5 (tried using 1.42.0)

**Comment by brandur** [comment]
@keithblaha Thanks for the repro! That definitely helps me a lot.

Okay, I see what's going on here. The auto-paginator essentially relies on passing parameters and options off to new `StripeCollection` instances that are created in order to produce a chain of successful auths. The problem here is that subcollections (e.g. invoice items) that are instantiated below a parent resource item (e.g. invoices) are not handed new parameters or options and therefore cannot paginate. I'm taking a look now to see what we can do about it.

**Comment by keithblaha** [comment]
:+1: awesome, thanks for being so responsive!

**Comment by brandur** [comment]
I thought about this one a little more, and I'm now a little worried that I may have made a mistake when designing the original interface here :) 

Because most of the collection-returning methods were already performing real fetches, I patched the iterator helper (`.autoPagingIterable()`) onto the already-materialized list objects. What it should probably have looked like instead is something a little more explicit that gets you a proxy object until you specificy credentials:

``` java
Invoice.list().autoPagingIterable(paginationOptions, requestOptions)
```

Instead of:

``` java
Invoice.list(paginationOptions, requestOptions).autoPagingIterable()
```

Then we could have consistent helpers on subcollections like:

``` java
invoice.getLines().autoPagingIterable(paginationOptions, requestOptions)
```

Everything would be explicit, and there would be no expectation of params/options implicitly making their way through objects.

Because the interface is not designed this way, I figure we have two options here.

1. Keep the current interface, but allow params/options to be passed to `.autoPagingIterable()`. This new version would then be recommended for use with all subcollections. The advantage of this approach is that it's quite easy. The disadvantage is we'd be allowing the implicit passing of params/options in some places, but requiring explicit passing in others.
2. Modify the `LiveStripeResponseGetter` so that after it performs a Gson deserialization, it inspects the object via reflection and tweaks any internal `StripeCollection` instances it sees to be given the same params/options as were used with the original request. The advantage of this approach is user convenience. The disadvantages are (1) increased complexity and taking us further down this rabbit hole of passing implicit options around all over the place, and (2) some additional performance impact from the second reflection step.

I'd definitely appreciate some feedback on what we should try here!

/cc @jimdanz @russelldavis @keithblaha 
/cc @stripe/api-libraries

**Comment by brandur** [comment]
And for what it's worth, this is almost certaily broken in other languages too (e.g. Ruby). We may just have gotten lucky so far in that many people are configuring credentials at the global level.

**Comment by keithblaha** [comment]
I'm personally fine with either solution, however faster turn around time and lower complexity of the first proposal are attractive

**Comment by keithblaha** [comment]
Hey @brandur, I was just wondering if there has been any development on this issue yet

**Comment by brandur** [comment]
@keithblaha I've added a proposed changed over in #269 that get this patched. Let me know what you think about the new interface.

**Comment by keithblaha** [comment]
Thanks @brandur, the new interface looks good to me. Is there an ETA on merging?

**Comment by brandur** [comment]
Tried another pass on solving this problem in #275. It's not quite as elegant, but continues to support everything that needs to be supported.

**Comment by keithblaha** [comment]
Looks good to me. Have you tested against the gist? https://gist.github.com/keithblaha/018594d285b1568792d5

**Comment by brandur** [comment]
@keithblaha It requires that this line:

``` java
for (InvoiceLineItem line : invoice.getLines().autoPagingIterable()) {
```

Be change to this:

``` java
for (InvoiceLineItem line : invoice.getLines().autoPagingIterable(null, requestOptions)) {
```

But yep, I just tried it, and it works.

**Comment by keithblaha** [comment]
Awesome, thank you! :+1: 

I'll be watching the PR for it's merge/release
